pkg/util/atomic_copier.go

```diff
@@ -0,0 +1,132 @@
+package util
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"os"
+	"os/exec"
+	"strings"
+
+	"k8s.io/klog/v2"
+)
+
+var (
+	cpArgs = []string{
+		"--verbose",
+		"--recursive",
+		"--preserve",
+		"--reflink=auto",
+	}
+)
+
+// AtomicCopier performs a two operation: copies source path to an intermediate location,
+// then renames from intermediate to final destination.
+// On Unix systems, the rename operation is atomic. This ensures that the file
+// is either fully renamed or not at all, which helps prevent issues like partial
+// file copies in cases of power failure or unexpected program termination.
+type AtomicCopier struct {
+	Source      string
+	Destination string
+
+	intermediatePath string
+}
+
+func (c *AtomicCopier) CopyToIntermediate() error {
+	if c == nil {
+		return nil
+	}
+	c.intermediatePath = fmt.Sprintf("%s.tmp", c.Destination)
+	if exists, err := PathExists(c.intermediatePath); err != nil {
+		return err
+	} else if exists {
+		if err := os.RemoveAll(c.intermediatePath); err != nil {
+			klog.Errorf("Failed to remove intermediate path %q which already existed: %v", c.intermediatePath, err)
+			return fmt.Errorf("failed to remove %q: %w", c.intermediatePath, err)
+		}
+	}
+	cmd := exec.Command("cp", append(cpArgs, c.Source, c.intermediatePath)...) //nolint:gosec
+
+	var outb, errb bytes.Buffer
+	cmd.Stdout = &outb
+	cmd.Stderr = &errb
+	err := cmd.Run()
+
+	if err != nil {
+		klog.ErrorS(nil, "Failed to make an intermediate copy", "cmd", cmd,
+			"stdout", strings.ReplaceAll(outb.String(), "\n", `, `),
+			"stderr", errb.String())
+		copyErr := fmt.Errorf("failed to copy %q to %q: %w", c.Source, c.intermediatePath, err)
+		if err := os.RemoveAll(c.intermediatePath); err != nil {
+			return errors.Join(copyErr, fmt.Errorf("failed to remove intermediate path %q: %w", c.intermediatePath, err))
+		}
+		return copyErr
+	}
+	klog.InfoS("Made an intermediate copy", "cmd", cmd)
```

Comment:
Do you mean this?
```suggestion
	klog.InfoS("Completed an atomic file copy operation", "cmd", cmd)
```

Reply:
Not really, not yet at least. The "atomic file copy operation" will be complete after rename, it's just a first part.